presto-spark-base/src/main/java/com/facebook/presto/spark/execution/NativeExecutionTask.java

```diff
@@ -136,138 +106,112 @@ public NativeExecutionTask(
                     .keySet().stream()
                     .map(outputId -> new BufferInfo(outputId, false, 0, 0, PageBufferInfo.empty()))
                     .collect(toImmutableList());
-            this.taskInfo.set(createInitialTask(taskId, URI.create(LOCAL_URI + taskId), bufferStates, new TaskStats(DateTime.now(), null), ""));
-            this.taskUpdateRequestCodec = taskUpdateRequestCodec;
-            this.planFragmentCodec = planFragmentCodec;
-            this.taskInfoCodec = taskInfoCodec;
+            this.workerClient = new PrestoSparkHttpWorkerClient(httpClient, taskId, URI.create(LOCAL_URI));
+            this.taskInfoFetcher = new HttpNativeExecutionTaskInfoFetcher(
+                    this.updateScheduledExecutor,
+                    this.workerClient,
+                    this.executor,
+                    this.taskId);
+            this.taskResultFetcher = new HttpNativeExecutionTaskResultFetcher(
+                    this.updateScheduledExecutor,
+                    this.workerClient,
+                    this.taskId,
+                    Optional.empty());
         }
     }
 
+    /**
+     * Gets the most updated TaskInfo of the task of the native task.
+     *
+     * @return most updated TaskInfo, null if TaskInfoFetcher has not yet retrieved the very first TaskInfo.
+     */
     public TaskInfo getTaskInfo()
     {
-        return taskInfo.get();
+        return taskInfoFetcher.getTaskInfo();
     }
 
-    public void start()
+    /**
+     * Blocking call to poll from result fetcher buffer. Blocks until content becomes available in the buffer, or until timeout is hit.
+     *
+     * @return the first SerializedPage result fetcher buffer contains.
+     */
+    public SerializedPage pollResult()
```

Comment:
Better to make the return value as `Optional<SerializedPage>` to indicate it could be null?

Reply:
Good point!